File: 0011_container_with_most_water/solution.py
# LeetCode 0011 - Container With Most Water
# https://leetcode.com/problems/container-with-most-water/

from typing import List


class Solution:
    def maxArea(self, height: List[int]) -> int:
        left = 0
        right = len(height) - 1
        best = 0

        while left < right:
            width = right - left
            best = max(best, min(height[left], height[right]) * width)
            if height[left] < height[right]:
                left += 1
            else:
                right -= 1

        return best

File: 0011_container_with_most_water/solution.go
// LeetCode 0011 - Container With Most Water
// https://leetcode.com/problems/container-with-most-water/

func maxArea(height []int) int {
	left, right := 0, len(height)-1
	best := 0

	for left < right {
		width := right - left
		h := height[left]
		if height[right] < h {
			h = height[right]
		}
		if h*width > best {
			best = h * width
		}
		if height[left] < height[right] {
			left++
		} else {
			right--
		}
	}

	return best
}
